### PR TITLE
First steps toward Python 2 / 3 compatibility

### PR DESCRIPTION
**Pull request by ChristopherChudzicki**

### CODE REVIEW

**Comment by jolyonb** [comment]
Can you add this statement to the top of basically every nonempty file?

`from __future__ import division, absolute_import, unicode_literals`

I don't think we need `print_function`, as we should never be using a `print` statement.

File: six.py [hunk 1]
```diff
@@ -0,0 +1,953 @@
+# Copyright (c) 2010-2019 Benjamin Peterson
```
Comment (jolyonb) [inline six.py]: Wait, why is this included? `six` should be included in `requirements.txt` and installed via `pip`.

Comment (ChristopherChudzicki) [reply six.py]: Do we know if `edx` will have their own copy of `six` available? If not, we'll need to include it like we do for `voluptuous`, right?

Comment (jolyonb) [reply six.py]: You know what? I'm going to ask the edX people right now...

Comment (jolyonb) [reply six.py]: Actually, I just checked in studio - six is presently available to codejail as a standard python library. I assume it will be in python3 as well (if not, I'll raise hell about it).

Comment (ChristopherChudzicki) [reply six.py]: Ah, smart, I should have checked. Happy to not include it, then :)

**Comment by ChristopherChudzicki** [comment]
> Can you add this statement to the top of basically every nonempty file?

> `from __future__ import division, absolute_import, unicode_literals`

If we do this, I think we'll need to use `six.text_type` in all of our voluptuous schema. I guess that's the point though...

**Comment by jolyonb** [comment]
I think that's exactly the point, alas...

**Comment by jolyonb** [comment]
I'm happy to review and merge as-is; it may not yet work for python3, but so long as it continues to work for python2, we can chip away at the required improvements. What do you think?

File: mitxgraders/comparers/__init__.py [hunk 2]
```diff
@@ -1,6 +1,8 @@
+from __future__ import print_function, division, absolute_import
 # Comparers can be imported by using "from mitxgraders import comparers"
```
Comment (jolyonb) [inline mitxgraders/comparers/__init__.py]: This comment seems out-dated now?

File: requirements3.txt [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Used by the library (versions most recent, except six mirrors edX current)
+numpy==1.16.4
+scipy==1.2.2
+pyparsing==2.0.7
```
Comment (jolyonb) [inline requirements3.txt]: pyparsing is at 2.4.0, I think?

File: tests/formulagrader/test_formulagrader.py [hunk 4]
```diff
@@ -215,8 +215,9 @@ def test_fg_userfunction():
         )
         grader(None, "that'sbad(1)")
 
-    expect = ("1 is not a valid key, must be of <type 'str'> for dictionary "
-              "value @ data\['user_functions'\]. Got {1: <ufunc 'tan'>}")
+    expect = ("1 is not a valid key, must be of {str_type} for dictionary "
+              "value @ data\['user_functions'\]. Got {{1: <ufunc 'tan'>}}").format(
+              str_type=str(str))
```
Comment (jolyonb) [inline tests/formulagrader/test_formulagrader.py]: Did you mean `type(str)` here?

File: tests/formulagrader/test_formulagrader.py [hunk 5]
```diff
@@ -231,8 +232,8 @@ def test_fg_userconstants():
     )
     assert grader(None, "hello")['ok']
 
-    expect = ("1 is not a valid key, must be of <type 'str'> for dictionary "
-              "value @ data\['user_constants'\]. Got {1: 5}")
+    expect = ("1 is not a valid key, must be of {str_type} for dictionary "
+              "value @ data\['user_constants'\]. Got {{1: 5}}").format(str_type=str(str))
```
Comment (jolyonb) [inline tests/formulagrader/test_formulagrader.py]: `type(str)`?

File: tests/helpers.py [hunk 6]
```diff
@@ -33,3 +36,28 @@ def _func(*args, **kwargs):
...
+    """
+    pattern = "([0-9]*\.[0-9]{{{round_to}}}[0-9]*)".format(round_to=round_to)
+    def replacer(match):
+        number = float(match.group(1))
+        formatter = "{{0:.{round_to}f}}".format(round_to=round_to)
+        return formatter.format(number)
+
+    return re.sub(pattern, replacer, string)
+
+if sys.version_info >= (3,):
```
Comment (jolyonb) [inline tests/helpers.py]: I'd prefer to do this like follows:
```
try:
    import unittest.mock as mock
except NameError:
    import mock
```

File: tests/helpers.py [hunk 7]
```diff
@@ -33,3 +36,28 @@ def _func(*args, **kwargs):
...
+
+    Usage
+    =====
+    >>> s = "pi is 3.141592653589793 and e is 2.71828182845904523536028747 and one is 1.000"
+    >>> round_decimals_in_string(s)
+    'pi is 3.141593 and e is 2.718282 and one is 1.000'
+
+    Note that the final occurrence of 1.000 was not rounded.
+    """
+    pattern = "([0-9]*\.[0-9]{{{round_to}}}[0-9]*)".format(round_to=round_to)
```
Comment (jolyonb) [inline tests/helpers.py]: First star doesn't need to be there. Second star should be a +.

File: tests/helpers/calc/test_math_array.py [hunk 8]
```diff
@@ -126,7 +127,7 @@ def test_addition_with_other_types():
     with raises(ShapeError, match=match):
         A + (1 + 2j)
     with raises(TypeError, match="Cannot add/subtract a matrix with object of "
-                "<type 'list'>"):
+                "{list_type}".format(list_type=list)):
```
Comment (jolyonb) [inline tests/helpers/calc/test_math_array.py]: Did you mean `type(list)`?

File: tests/helpers/calc/test_math_array.py [hunk 9]
```diff
@@ -178,7 +179,7 @@ def test_subtraction_with_other_types():
     with raises(ShapeError, match=match):
         A - (1 + 2j)
     with raises(TypeError, match="Cannot add/subtract a matrix with object of "
-                "<type 'list'>"):
+                "{list_type}".format(list_type=list)):
```
Comment (jolyonb) [inline tests/helpers/calc/test_math_array.py]: `type(list)`?

File: tests/helpers/calc/test_math_array.py [hunk 10]
```diff
@@ -297,11 +298,11 @@ def test_tensor_multiplication_not_supported():
 def test_multiplication_with_unexpected_types_raises_TypeError():
     A = random_math_array([4, 2])
     B = [1, 2, 3]
-    match = "Cannot multiply a matrix with object of <type 'list'>"
+    match = "Cannot multiply a matrix with object of {list_type}".format(list_type=list)
```
Comment (jolyonb) [inline tests/helpers/calc/test_math_array.py]: `type(list)`?

File: tests/helpers/calc/test_math_array.py [hunk 11]
```diff
@@ -297,11 +298,11 @@ def test_tensor_multiplication_not_supported():
...
     with raises(TypeError, match=match):
         A*B
 
-    match = "Cannot multiply object of type <type 'list'> with a matrix."
+    match = "Cannot multiply object of type {list_type} with a matrix.".format(list_type=list)
```
Comment (jolyonb) [inline tests/helpers/calc/test_math_array.py]: `type(list)`?

File: tests/helpers/calc/test_math_array.py [hunk 12]
```diff
@@ -324,7 +325,7 @@ def test_division_by_array_raises_error():
     with raises(ShapeError, match=match):
         A/A
 
-    match = "Cannot divide vector by object of <type 'list'>"
+    match = "Cannot divide vector by object of {list_type}".format(list_type=list)
```
Comment (jolyonb) [inline tests/helpers/calc/test_math_array.py]: `type(list)`?

File: tests/helpers/calc/test_math_array.py [hunk 13]
```diff
@@ -369,7 +370,7 @@ def test_power_error_messages():
         2**A
 
     # Weird case with MathArray on the right.
-    match = "Cannot raise <type 'list'> to power of matrix."
+    match = "Cannot raise {list_type} to power of matrix.".format(list_type=list)
```
Comment (jolyonb) [inline tests/helpers/calc/test_math_array.py]: `type(list)`?

File: tests/helpers/calc/test_math_array.py [hunk 14]
```diff
@@ -393,7 +394,7 @@ def test_power_error_messages():
     with raises(ShapeError, match=match):
         A**B
 
-    match = "Cannot raise matrix to power of type <type 'list'>."
+    match = "Cannot raise matrix to power of {list_type}.".format(list_type=list)
```
Comment (jolyonb) [inline tests/helpers/calc/test_math_array.py]: `type(list)`?

File: tests/helpers/calc/test_math_array.py [hunk 15]
```diff
@@ -414,7 +415,7 @@ def test_scalar_special_cases():
     assert five**three == 125
 
     # With MathArrays on the right:
-    match = "Cannot raise a scalar to a <type 'list'> power"
+    match = "Cannot raise a scalar to a {list} power".format(list=list)
```
Comment (jolyonb) [inline tests/helpers/calc/test_math_array.py]: `type(list)`?

File: tests/test_listgrader.py [hunk 16]
```diff
@@ -851,7 +852,7 @@ def test_errors():
         grader(None, ["Hello"])
 
     # Bad input
-    msg = "Expected answer to have type <type list>, but received <type 'tuple'>"
+    msg = "Expected answer to have {list}, but received {tuple}".format(list=list, tuple=tuple)
```
Comment (jolyonb) [inline tests/test_listgrader.py]: `type(list)`? `type(tuple)`?

File: tests/test_base.py [hunk 17]
```diff
@@ -112,7 +113,7 @@ def test_itemgrader():
     with raises(ConfigError, match="Expected at least one answer in answers"):
         grader = StringGrader(answers=())
         grader(None, "hello")
-    with raises(ConfigError, match="Expected string for student_input, received <type 'int'>"):
+    with raises(ConfigError, match="Expected string for student_input, received {int}".format(int=int)):
```
Comment (jolyonb) [inline tests/test_base.py]: `type(int)`? Perhaps I'm getting these all confused, but this code seems weird to me...

**Comment by jolyonb** [comment]
I've reviewed everything, left a bunch of comments. Let's clear these all up and merge, so the next step is self-contained.

File: requirements3.txt (hunk 3 above)
Comment (ChristopherChudzicki) [reply requirements3.txt]: fixed, pip-installed, and not causing any new breaks in python 3 testing

File: tests/helpers.py (hunk 7 above)
Comment (ChristopherChudzicki) [reply tests/helpers.py]: 👍

File: tests/helpers.py (hunk 6 above)
Comment (ChristopherChudzicki) [reply tests/helpers.py]: turns out it's ImportError :)

File: tests/formulagrader/test_formulagrader.py (hunk 4 above)
Comment (ChristopherChudzicki) [reply tests/formulagrader/test_formulagrader.py]: I'm changing this to `str_type=str`, see below.

File: tests/formulagrader/test_formulagrader.py (hunk 5 above)
Comment (ChristopherChudzicki) [reply tests/formulagrader/test_formulagrader.py]: Changing this to `str_type=str`, also.

File: tests/helpers/calc/test_math_array.py (hunk 8 above)
Comment (ChristopherChudzicki) [reply tests/helpers/calc/test_math_array.py]: No, because `type(list)` is `<type 'type'>` in Python 2 and `<class 'type'>` in Python 3. We want the type of the instance (name,y, `list`), not the type of the class/type.

Also works: `.format(type_list=type([])`, but it annoyed me to instantiate a list for no reason, plus I found `.format(type_tuple=type(tuple()))` to be ugly.

**Comment by ChristopherChudzicki** [comment]
Ready for re-review

File: tests/helpers/calc/test_math_array.py (hunk 8 above)
Comment (jolyonb) [reply tests/helpers/calc/test_math_array.py]: :+1:

File: mitxgraders/listgrader.py [hunk 18]
```diff
@@ -373,8 +374,8 @@ def check(self, answers, student_input, **kwargs):
             results = [self.perform_check(answer_list, student_input) for answer_list in answers]
             return self.get_best_result(results)
         else:
-            msg = "Expected answer to have type <type list>, but received {}"
-            raise ConfigError(msg.format(type(student_input)))
+            msg = "Expected answer to have {0}, but received {1}"
+            raise ConfigError(msg.format(type(list()), type(student_input)))
```
Comment (jolyonb) [inline mitxgraders/listgrader.py]: `type(list())` -> `list`?

File: tests/formulagrader/test_formulagrader.py (hunk 5 above)
Comment (jolyonb) [reply tests/formulagrader/test_formulagrader.py]: I think you missed this one.

**Comment by jolyonb** [comment]
Re-reviewed. Two tiny outstanding issues, and then this is ready for merge.

File: mitxgraders/listgrader.py (hunk 18 above)
Comment (ChristopherChudzicki) [reply mitxgraders/listgrader.py]: A remnant from when I was confused 😄

**Comment by ChristopherChudzicki** [comment]
Ready
